zcash_client_sqlite/src/wallet/init.rs

```diff
@@ -411,6 +411,87 @@ impl<P: consensus::Parameters> RusqliteMigration for WalletMigration2<P> {
+    "Add views over transaction & note data."
+);
+
+impl RusqliteMigration for WalletMigrationAddTxViews {
+    type Error = WalletMigrationError;
+
+    fn up(&self, transaction: &Transaction) -> Result<(), WalletMigrationError> {
+        transaction.execute_batch(
+            "CREATE VIEW IF NOT EXISTS v_tx_sent_notes AS
+            SELECT transactions.id_tx         AS id,
+                   transactions.block         AS mined_height,
+                   transactions.tx_index      AS transaction_index,
+                   transactions.txid          AS raw_transaction_id,
+                   transactions.expiry_height AS expiry_height,
+                   transactions.raw           AS raw,
+                   sent_notes.address         AS to_address,
+                   sent_notes.value           AS value,
+                   sent_notes.memo            AS memo,
+                   sent_notes.id_note         AS note_id,
+                   blocks.time                AS block_time
+            FROM   transactions
+                   LEFT JOIN sent_notes
+                          ON transactions.id_tx = sent_notes.tx
+                   LEFT JOIN blocks
+                          ON transactions.block = blocks.height;
+            CREATE VIEW IF NOT EXISTS v_tx_received_notes AS
+            SELECT transactions.id_tx     AS id,
+                   transactions.block     AS mined_height,
+                   transactions.tx_index  AS transaction_index,
+                   transactions.txid      AS raw_transaction_id,
+                   received_notes.value   AS value,
+                   received_notes.memo    AS memo,
+                   received_notes.id_note AS note_id,
+                   blocks.time            AS block_time,
+                   IFNULL(received_notes.is_change, false) as is_change
+            FROM   transactions
+                   LEFT JOIN received_notes
+                          ON transactions.id_tx = received_notes.tx
+                   LEFT JOIN blocks
+                          ON transactions.block = blocks.height;
+            CREATE VIEW IF NOT EXISTS v_tx_notes AS
+            SELECT transactions.id_tx           AS id,
+                   transactions.block           AS mined_height,
+                   transactions.tx_index        AS transaction_index,
+                   transactions.txid            AS raw_transaction_id,
+                   transactions.expiry_height   AS expiry_height,
+                   transactions.raw             AS raw,
+                   sent_notes.address           AS to_address,
+                   IFNULL(sent_notes.value, received_notes.value) AS value,
+                   IFNULL(sent_notes.memo, received_notes.memo)   AS memo,
+                   IFNULL(sent_notes.id_note, received_notes.id_note) AS note_id,
+                   blocks.time                  AS block_time,
+                   IFNULL(received_notes.is_change, false) as is_change
```

Comment:
AFAICT this is going to cause received notes to never show up in this view, when a sent note is also present.